### PR TITLE
fix: annotation preparation list iteration bug

### PR DESCRIPTION
**Pull request by adamliter**
## What does this PR address?

Fixes bug reported in [Slack](https://bentoml.slack.com/archives/CKRANBHPH/p1714672615101629).

## Before submitting:

- [x] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [x] Does the code follow BentoML's code style, `pre-commit run -a` script has passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [x] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [x] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: src/_bentoml_sdk/_pydantic.py [hunk 1]
```diff
@@ -53,13 +53,14 @@ def numpy_prepare_pydantic_annotations(
     _, remaining_annotations = _known_annotated_metadata.collect_known_metadata(
         annotations
     )
-    for i, annotation in enumerate(remaining_annotations[:]):
+    orig_len = len(remaining_annotations)
+    for i, annotation in enumerate(reversed(remaining_annotations[:])):
```
Comment (frostming) [inline src/_bentoml_sdk/_pydantic.py:57]:
```suggestion
    for i, annotation in reversed(enumerate(remaining_annotations[:])):
```

So you don't need to calculate the actual index

File: src/_bentoml_sdk/_pydantic.py [hunk 2]
```diff
@@ -80,13 +81,14 @@ def torch_prepare_pydantic_annotations(
...
```
Comment (frostming) [inline src/_bentoml_sdk/_pydantic.py:85]: Ditto

File: src/_bentoml_sdk/_pydantic.py [hunk 3]
```diff
@@ -106,13 +108,14 @@ def tf_prepare_pydantic_annotations(
...
```
Comment (frostming) [inline src/_bentoml_sdk/_pydantic.py:112]: Ditto

File: src/_bentoml_sdk/_pydantic.py [hunk 4]
```diff
@@ -168,11 +171,12 @@ def pathlib_prepare_pydantic_annotations(
     _, remaining_annotations = _known_annotated_metadata.collect_known_metadata(
         annotations
     )
+    orig_len = len(remaining_annotations)
     content_type: str | None = None
-    for i, annotation in enumerate(remaining_annotations[:]):
+    for i, annotation in enumerate(reversed(remaining_annotations[:])):
```
Comment (frostming) [inline src/_bentoml_sdk/_pydantic.py:176]: Ditto

**Review by frostming — CHANGES_REQUESTED** [review]

File: src/_bentoml_sdk/_pydantic.py (hunk 1 above)
Comment (adamliter) [reply src/_bentoml_sdk/_pydantic.py:57]: @frostming Does that work for you? I get a `TypeError` with Python 3.9:

```python
>>> my_list = ["a", "b"]
>>> for i, e in reversed(enumerate(my_list[:])):
...     print(i, e)
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'enumerate' object is not reversible
```

Comment (adamliter) [reply src/_bentoml_sdk/_pydantic.py:57]: It looks like this works:

```python
>>> my_list = ["a", "b"]
>>> for i, e in reversed(list(enumerate(my_list[:]))):
...     print(i, e)
... 
1 b
0 a
```

Comment (frostming) [reply src/_bentoml_sdk/_pydantic.py]: Fine, but `[:]` is no necessary then, can you remove it?

Comment (adamliter) [reply src/_bentoml_sdk/_pydantic.py]: Yes, good catch. Thanks for the feedback. I've removed it.

**Review by frostming — APPROVED** [review]

**Pull request merged**